File: Source/AlphaTab/Rendering/Effects/NoteEffectInfoBase.cs
using AlphaTab.Collections;
using AlphaTab.Model;
using AlphaTab.Rendering.Glyphs;

namespace AlphaTab.Rendering.Effects
{
    abstract class NoteEffectInfoBase : IEffectBarRendererInfo
    {
        protected FastList<Note> LastCreateInfo;

        public virtual bool ShouldCreateGlyph(Settings settings, Beat beat)
        {
            LastCreateInfo = new FastList<Note>();
            for (int i = 0, j = beat.Notes.Count; i < j; i++)
            {
                var n = beat.Notes[i];
                if (ShouldCreateGlyphForNote(n))
                {
                    LastCreateInfo.Add(n);
                }
            }
            return LastCreateInfo.Count > 0;
        }

        protected abstract bool ShouldCreateGlyphForNote(Note note);

        public abstract string EffectId { get; }
        public virtual bool HideOnMultiTrack { get { return false; } }
        public virtual bool CanShareBand { get { return true; } }
        public abstract EffectBarGlyphSizing SizingMode { get; }
        public abstract EffectGlyph CreateNewGlyph(BarRendererBase renderer, Beat beat);

        public virtual bool CanExpand(Beat @from, Beat to)
        {
            return true;
        }
    }
}

File: src/rendering/effects/NoteEffectInfoBase.ts
import { Beat } from '@src/model/Beat';
import { Note } from '@src/model/Note';
import { EffectBarRendererInfo } from '@src/rendering/EffectBarRendererInfo';
import { Settings } from '@src/Settings';

export abstract class NoteEffectInfoBase extends EffectBarRendererInfo {
    protected lastCreateInfo: Note[] | null = null;

    public shouldCreateGlyph(settings: Settings, beat: Beat): boolean {
        this.lastCreateInfo = [];
        for (let i: number = 0, j: number = beat.notes.length; i < j; i++) {
            let n: Note = beat.notes[i];
            if (this.shouldCreateGlyphForNote(n)) {
                this.lastCreateInfo.push(n);
            }
        }
        return this.lastCreateInfo.length > 0;
    }

    protected abstract shouldCreateGlyphForNote(note: Note): boolean;

    public get hideOnMultiTrack(): boolean {
        return false;
    }

    public get canShareBand(): boolean {
        return true;
    }

    public canExpand(from: Beat, to: Beat): boolean {
        return true;
    }
}
